### PR TITLE
[9.0] MIG bi_sql_editor

### PR DESCRIPTION
**Pull request by nicomacr**

### CODE REVIEW

**Comment by AaronHForgeFlow** [comment]
I :heart: this module. Would you include the changes proposed here https://github.com/OCA/reporting-engine/pull/149 ?

**Comment by JordiBForgeFlow** [comment]
@aheficent perhaps you can propose a PR that includes it

**Comment by legalsylvain** [comment]
please wait, i'm doing some improvement on #149.

**Comment by legalsylvain** [comment]
improvment available here : https://github.com/OCA/reporting-engine/pull/150

**Comment by legalsylvain** [comment]
Hi @aheficent. Thanks a lot for porting this module. LGTM for me (code / test). please ping me once you integrated #150 commit.
regards.

**Comment by AaronHForgeFlow** [comment]
Sure @legalsylvain Thank you for the improvement

**Comment by JordiBForgeFlow** [comment]
@nicomacr @legalsylvain @aheficent I have improved 9.0 with the cherry pick of #150 here: https://github.com/adhoc-dev/reporting-engine-1/pull/1

**Comment by nicomacr** [comment]
@jbeficent thanks for the PR

File: bi_sql_editor/demo/bi_sql_view.xml
Comment (legalsylvain) [inline bi_sql_editor/demo/bi_sql_view.xml]: because, odoo renamed authorship in V9.

    case
        when author ilike '%OpenERP SA%' THEN 'Odoo SA'
        when author ilike '%Odoo SA%' THEN 'Odoo SA'
        when author ilike '%Odoo Community Association (OCA)%' THEN 'OCA'
       else 'Undefined Author' END as x_author_type

**Comment by AaronHForgeFlow** [comment]
Hi @nicomacr I'm proposing some tests here: https://github.com/adhoc-dev/reporting-engine-1/pull/2 . Also it fixes the unlink method, which is not calling super.

**Review by legalsylvain — APPROVED** [review]
LGTM to me. one comment on demo data to update.

Otherwise, :+1: 
Thanks ! 

PS : could you review, the V8.0 PR ? 
https://github.com/OCA/reporting-engine/pull/150

**Comment by AaronHForgeFlow** [comment]
I closed the previous PR for the test. I created a new one here: https://github.com/adhoc-dev/reporting-engine-1/pull/3 but it's failing because it's creating a lot of connections to the database. I'll update the PR when this issue is solved

**Comment by legalsylvain** [comment]
Any other review to validate the PR ? thanks !

**Comment by JordiBForgeFlow** [comment]
Can we add an option for the person thay creates the query to decide the order of the.views? (Treen graph, pivot)?

**Comment by AaronHForgeFlow** [comment]
@legalsylvain This works ok to me. However there's must be something wrong with the test I did because when I run it on local it's trying to connect to the database multiple times. Getting this:
PoolError: The Connection Pool Is Full

Honestly, don't know the reason

File: bi_sql_editor/README.rst
Comment (elicoidal) [inline bi_sql_editor/README.rst]: add a blank line

File: bi_sql_editor/README.rst
Comment (elicoidal) [inline bi_sql_editor/README.rst]: Best would be enumerated list (nice to have)

File: bi_sql_editor/__openerp__.py
Comment (elicoidal) [inline bi_sql_editor/__openerp__.py]: demo/res_groups_demo.xml (several)

File: bi_sql_editor/demo/res_groups.xml
Comment (elicoidal) [inline bi_sql_editor/demo/res_groups.xml]: odoo
+Improve the format

File: bi_sql_editor/models/bi_sql_view.py
Comment (elicoidal) [inline bi_sql_editor/models/bi_sql_view.py]: remove (

File: bi_sql_editor/models/bi_sql_view.py
Comment (elicoidal) [inline bi_sql_editor/models/bi_sql_view.py]: Syntax should follow: https://www.postgresql.org/"
        "docs/current/static/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIERS

File: bi_sql_editor/models/bi_sql_view_field.py
Comment (elicoidal) [inline bi_sql_editor/models/bi_sql_view_field.py]: s/let/keep

File: bi_sql_editor/models/bi_sql_view_field.py
Comment (elicoidal) [inline bi_sql_editor/models/bi_sql_view_field.py]: Comodel

**Review by elicoidal — CHANGES_REQUESTED** [review]
nice one

**Comment by legalsylvain** [comment]
> Can we add an option for the person that creates the query to decide the order of the.views? (Treen graph, pivot)?

We could 
- add an extra char field in the view, with a default value to 'graph,pivot,tree', (readonly if sql is not in the draft / analyzed state)
- add a constraint that check that value.split(',') contains only ['graph', 'pivot', 'tree']
- use this value to generate the action. (and use it in the button_open_ui)

No time for this feature, for my side, but ok to review it. (It could be done in the second PR.)

**Comment by AaronHForgeFlow** [comment]
@legalsylvain I'll propose a PR for that

**Comment by AaronHForgeFlow** [comment]
Hi @nicomacr @legalsylvain I added an option for the person that creates the query to decide the first view to show: https://github.com/adhoc-dev/reporting-engine-1/pull/4

**Review by LoisRForgeFlow — APPROVED** [review]
I agree with @elicoidal remarks but apart from that this module is great!

Thank you for the great contribution.

**Comment by legalsylvain** [comment]
@nicomacr , @aheficent Could you make minor fixes asked by @elicoidal  ? 

:+1: Thanks a lot again. (The feature to select view types is a great improvement !)

regards.

File: bi_sql_editor/models/bi_sql_view.py
Comment (JordiBForgeFlow) [inline bi_sql_editor/models/bi_sql_view.py]: Please rename `Graph, action and Menu Created` to `Views, Action and Menu Created`, because now you show pivot and tree as well.

**Comment by AaronHForgeFlow** [comment]
Hi @legalsylvain @nicomacr I just created a new Pull Request for the fixes proposed here. Can you review? Thanks https://github.com/adhoc-dev/reporting-engine-1/pull/5

**Comment by legalsylvain** [comment]
2 bugs on CI.
1. ir.model.fields.create() includes unknown fields: tree_visibility
2. No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself. -> something wrong happened. don't know why.

**Review by elicoidal — APPROVED** [review]

**Review by Emiq2 — APPROVED** [review]

**Comment by JordiBForgeFlow** [comment]
@pedrobaeza @gurneyalex do you know why would Travis stall on this one?

**Comment by AaronHForgeFlow** [comment]
@jbeficent @legalsylvain I think it's something related the tests. When I run those on local I got the following: `PoolError: The Connection Pool Is Full`

File: bi_sql_editor/models/bi_sql_view.py
Comment (AaronHForgeFlow) [inline bi_sql_editor/models/bi_sql_view.py:221]: It seems to me that the bi_sql records created in a test are never deleted. And when the bi_Sql record is validaded never goes back to draft. Besides, this line is giving errors when calling this method from a test. 

@astirpe Maybe you can shed some light here

BTW: An uninstall hook is needed to drop some materialized views and some relations

File: bi_sql_editor/demo/bi_sql_view_demo.xml
Comment (AaronHForgeFlow) [inline bi_sql_editor/demo/bi_sql_view_demo.xml]: This is causing an error when uninstalling and installing the module:
`relation "x_bi_sql_view_modules_view" already exists`

I'm trying to do an uninstalling hook including this and some other clauses to delete remaining tables and views.

Comment (astirpe) [reply bi_sql_editor/demo/bi_sql_view_demo.xml]: Looks like the same problem that was occurring when uninstalling and installing the bi_view_editor: an hook script like [this one](https://github.com/OCA/reporting-engine/blob/9.0/bi_view_editor/hooks.py) would solve this issue

File: bi_sql_editor/models/bi_sql_view.py
Comment (astirpe) [reply bi_sql_editor/models/bi_sql_view.py:221]: Strange error. I can take a look at it next week.

Comment (astirpe) [reply bi_sql_editor/models/bi_sql_view.py:221]: Also some warnings are raised:
```
2017-07-25 16:16:59,706 172 WARNING openerp_test openerp.models: ir.model.fields.create() includes unknown fields: tree_visibility
2017-07-25 16:16:59,710 172 WARNING openerp_test openerp.models: ir.model.fields.create() includes unknown fields: tree_visibility
```

Comment (AaronHForgeFlow) [reply bi_sql_editor/models/bi_sql_view.py:221]: This is what happens when running the test several times. The records are never deleted. It seems I need to run some queries every time I run the tests:
image

Moreover this have happened after running the uninstall hook, which looks like:
` ```
   cr.execute("""
        select relname
        from pg_class
        where relname like 'x_bi_sql_view%' and relkind='m'
        """)
    for r in cr.fetchall():
        cr.execute("""
               DROP MATERIALIZED VIEW %s
            """ % r)
```

    cr.execute("""
        select relname
        from pg_class
        where relname like 'x_bi_sql_view%' and relkind='r'
        """)
    for r in cr.fetchall():
        cr.execute("""
               DROP TABLE %s
            """ % r)
    cr.execute("""
        select table_name from INFORMATION_SCHEMA.views
        where table_name like 'x_bi_sql%'""")

    # Drop not materialized views
    for v in cr.fetchall():
        cr.execute("""
               DROP VIEW %s
            """ % v)

    #Drop table if uninstalling went wrong
    cr.execute("""
        delete from ir_model_fields where model like 'bi.sql.view%';
        delete from ir_model_fields where model like 'bi_sql_%';
        delete from ir_model where model like 'x_bi_sql_view.%';
        """)`

Comment (AaronHForgeFlow) [reply bi_sql_editor/models/bi_sql_view.py:221]: I'll propose another Pull Request fixing this

**Comment by AaronHForgeFlow** [comment]
I proposed a new Pull Request here: https://github.com/adhoc-dev/reporting-engine-1/pull/6

@nicomacr Please take a look at this before it's merged.

To sum up, the Pull request does the following:

- Adds uninstall hook: Avoid errors when reinstalling.
- Adds checks when deleting records in the button_set_draft method
- For non materialized views was getting errors when deleting the views. Added check
- Deleted test. No matter what I tried. Records created on the test are never deleted and consequently the views, the tables and so on. Moreover, if the test refers to a record created with demo data, the system enters in a loop, and start creating connections to the database to create views, tables and so on. Insane. I tried to reproduced manually and this was not happening, so I think it is a problem with the test environment. To sum up, test are deleted.

File: bi_sql_editor/models/bi_sql_view.py
Comment (legalsylvain) [inline bi_sql_editor/models/bi_sql_view.py]: why did you added a test ? if rule_id is False, sql_view.r_le_id.unlink() make nothing.

File: bi_sql_editor/hooks.py
Comment (legalsylvain) [inline bi_sql_editor/hooks.py]: I don't understand all the lines below. 15+
for me button_set_draft() and unlink() are enought, to drop all the (materialized or not) views.

File: bi_sql_editor/hooks.py
Comment (legalsylvain) [inline bi_sql_editor/hooks.py]: better to do sql_view.model_id.unlink(), don't you think ?

File: bi_sql_editor/models/bi_sql_view.py
Comment (legalsylvain) [inline bi_sql_editor/models/bi_sql_view.py]: same here.

**Review by legalsylvain — CHANGES_REQUESTED** [review]
Hi thanks a lot for adding a hook. 
Some comments inline however.

thanks.

File: bi_sql_editor/hooks.py
Comment (AaronHForgeFlow) [reply bi_sql_editor/hooks.py]: I'll check that again without tests. But with them that was not enough, because it creates a lot of dirty data in the database. Thank you @legalsylvain

**Comment by AaronHForgeFlow** [comment]
BTW, travis compilation was not failing because of the test here but for a test in another module: test_report_qweb_pdf_watermark

**Comment by MiquelRForgeFlow** [comment]
No, Travis now is not failing by the pdf_watermark. It seems it is failing because it cannot install some external libraries. They should be added in Travis or something like this may be.

**Comment by AaronHForgeFlow** [comment]
@mreficent All right. I'll try to fix that

**Comment by MiquelRForgeFlow** [comment]
@aheficent sorry, yes, the problem is with `report_qweb_pdf_watermark`

File: bi_sql_editor/hooks.py
Comment (AaronHForgeFlow) [reply bi_sql_editor/hooks.py]: I think the same. This is supposed to be done by this line `rec.button_set_draft()` But for some reason it does not work properly. At the end it shows this error:

> Unable to delete 121@ir.model
and then:
> UserError: (u"The field 'bi.sql.view.field.bi_sql_view_id' cannot be removed because the field 'bi.sql.view.bi_sql_view_field_ids' depends on it.", None)

That's why I put all those queries.

**Comment by AaronHForgeFlow** [comment]
BTW among other things, Lint check fails because of the queries in the uninstall hook. But those queries are necessary IMO so I'd keep that

**Comment by AaronHForgeFlow** [comment]
Hi @legalsylvain I created another Pull Request here to attend your comments. Despite my previous comment I realized that those queries in the uninstall hook are not needed. However this error is still shown when uninstalling:
> Unable to delete 121@ir.model
and then:
> UserError: (u"The field 'bi.sql.view.field.bi_sql_view_id' cannot be removed because the field 'bi.sql.view.bi_sql_view_field_ids' depends on it.", None)

Nevertheless, no bi.sql.view and no bi.sql.view.field remains in the database when uninstalling. That's why I removed the queries from the uninstall hook.

**Comment by AaronHForgeFlow** [comment]
Created Pull Request here to fix flake8 error https://github.com/adhoc-dev/reporting-engine-1/pull/8

However this is still pending:
https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#no-sql-injection

Does anybody knows how to configure travis file to avoid this check? :pray:

**Comment by legalsylvain** [comment]
@aheficent : 

``self.env.cr.commit() # pylint: disable=invalid-commit``

Sample : https://github.com/OCA/server-tools/blob/10.0/database_cleanup/models/purge_columns.py#L51

regards;

**Comment by AaronHForgeFlow** [comment]
Thank you @legalsylvain! I added this to my last Pull Request. (https://github.com/adhoc-dev/reporting-engine-1/pull/8)

**Comment by legalsylvain** [comment]
@jbeficent , @aheficent could you set a little +1 on the PR against V8 you cherry picked ? 
https://github.com/OCA/reporting-engine/pull/150.

thanks !

**Comment by AaronHForgeFlow** [comment]
Runbot is compiled. It's showing red because of this:
`ParseError: "no such savepoint
" while parsing /home/odoo/build/OCA/reporting-engine/bi_sql_editor/demo/bi_sql_view_demo.xml:52, near
<function model="bi.sql.view" name="button_validate_sql_expression" eval="([ref('module_sql_view')])"/>`

Should we create a savepoint before calling the method  button_validate_sql_expression?

Edit: Run in local in debug mode. Same error as here travis is getting. In the second try it fails for different reason, the relations created for the demo data in the previous try were still there. Same problem as the one I was having with the tests.

**Review by ValentinoDefelice — APPROVED** [review]

**Comment by AaronHForgeFlow** [comment]
Hi @nicomacr I created a new PR here: https://github.com/adhoc-dev/reporting-engine-1/pull/9

This fixes the error related with the save point that was introduced in the last commit. Besides I added some tests back. However, the tests don't cover the part of creating the models because that would leave unwanted data in the database.
@legalsylvain @jbeficent

**Comment by AaronHForgeFlow** [comment]
@nicomacr Sorry for the Lint errors. :face_with_head_bandage:  Created new PR for that:https://github.com/adhoc-dev/reporting-engine-1/pull/10   Probably a commit squashing will be needed later

**Comment by AaronHForgeFlow** [comment]
Created PR to prevent runbot warning https://github.com/adhoc-dev/reporting-engine-1/pull/11

Travis will be fixed when this is merged: https://github.com/OCA/reporting-engine/pull/154

File: bi_sql_editor/models/bi_sql_view_field.py
Comment (legalsylvain) [inline bi_sql_editor/models/bi_sql_view_field.py]: thanks !

**Review by legalsylvain — APPROVED** [review]

**Review by JordiBForgeFlow — APPROVED** [review]
Technical and functional review :+1:

**Comment by JordiBForgeFlow** [comment]
Travis is KO due to problem found in #154. But tested locally, and all test are passing correctly.

The only problem we found is that it is not possible to include in the tests the parts where the models are created, because every time that you pass the test the tables created by the test are not deleted from the DB -- very strange --, so the next time you run the test, it fails because the table already exists.

**Comment by legalsylvain** [comment]
Well, I propose to merge this PR, because it's OK. (and it has 6 :+1:, that's a record ! )
let's go improving test in another PR, or for another version. (8.0 version didn't include test at all, so, it's better).

**Pull request merged**
